src/core/device.c

```diff
@@ -114,12 +156,129 @@ static void device_done(Unit *u) {
         assert(d);
 
         device_unset_sysfs(d);
+        free(d->devname);
         d->wants_property = strv_free(d->wants_property);
 }
 
+static int root_cgroup_apply_block_device_io_cost_qos(Unit *root_slice_unit, Device *d) {
+        char buf[DECIMAL_STR_MAX(dev_t)*2+9+(6+DECIMAL_STR_MAX(uint64_t)+1)*10+(6+DECIMAL_STR_MAX(loadavg_t)+1)*2];
+        IOCostQos *io_cost_qos;
+        dev_t dev;
+        bool enabled;
+        int r;
+
+        assert(root_slice_unit);
+        assert(d);
+
+        io_cost_qos = &d->io_cost_qos;
+
+        r = lookup_block_device(d->devname, &dev);
+        if (r < 0)
+                return r;
+
+        /* If ctrl is not set, the enabled value will be false */
+        enabled = io_cost_qos->ctrl != _IO_COST_CTRL_INVALID;
+
+        /* Will consider ctrl value of auto to be the same as not enabled.
+         * In both cases the kernel will automatically fill all other io_cost_qos
+         * parameteters so we do not need to input them */
+        if (io_cost_qos->ctrl == IO_COST_CTRL_AUTO || !enabled) {
+                log_info("IOCostQOSCtrl=auto or not set, setting all other values to kernel default");
+                xsprintf(buf, "%u:%u enable=%d ctrl=%.4s\n",
+                         major(dev),
+                         minor(dev),
+                         enabled,
+                         io_cost_ctrl_to_string(IO_COST_CTRL_AUTO));
+
+        } else if (io_cost_qos->ctrl == IO_COST_CTRL_USER) {
+                if (io_cost_qos->min > io_cost_qos->max)
+                         return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                                "IOCostQOSMin= is greater than IOCostQOSMax=, refusing to apply io.cost.qos settings");
+
+                xsprintf(buf, "%u:%u enable=%d ctrl=%.4s rpct=%u.%02u rlat=%u wpct=%u.%02u wlat=%u min=%lu.%02lu max=%lu.%02lu\n",
+                         major(dev),
+                         minor(dev),
+                         enabled,
+                         io_cost_ctrl_to_string(io_cost_qos->ctrl),
```

Comment:
Any reason why here the argument is `io_cost_qos->ctrl` while above it's const `IO_COST_CTRL_AUTO`? It looks inconsistent.

Reply:
when the value is `IO_COST_CTRL_AUTO` we do not bother setting any of the other fields. This argument could just be `IO_COST_CTRL_USER` though if that would be more consistent.